kubernetes/client/pod.go

```diff
@@ -0,0 +1,355 @@
+// command. If you want to pass StdIn or have a need for a reader / writer then
+// you'll need to use ExecCmd. If the command executes in the container and
+// stdErr is not nil, the command will return an error containing the contents
+// of stdErr.
+func (p *Kubernetesp) ExecInPod(podName string, containerName, namespace string, command []string) (string, string, error) {
+	stdOut := new(Writer)
+	stdErr := new(Writer)
+	opts := &CmdOptions{
+		Command: command,
+		StreamOptions: StreamOptions{
+			PodName:       podName,
+			ContainerName: containerName,
+			Out:           stdOut,
+			Err:           stdErr,
+		},
+	}
+
+	// Exec the command in the pod. If the exec call failed, return the error
+	_, _, err := p.ExecCmd(opts, namespace)
+	if err != nil {
+		return "", "", err
+	}
+
+	// If the exec call succeded, but the cmd failed, also error
+	if len(stdErr.Str) != 0 {
+		return strings.Join(stdOut.Str, " "), strings.Join(stdErr.Str, " "), fmt.Errorf("Command returned non zero: %s", stdErr.Str)
+	}
+
+	return strings.Join(stdOut.Str, " "), strings.Join(stdErr.Str, " "), nil
+
+}
+
+// PortForward sets up port forwarding to the manager that is running inside the
+// cluster.
+func (p *Kubernetesp) PortForward(pod *apiv1.Pod, requiredPorts []int) error {
+	if pod.Status.Phase != apiv1.PodRunning {
+		p.Error("unable to forward port because pod is not running.", "status", pod.Status.Phase, "pod", pod.ObjectMeta.Name)
+		return fmt.Errorf("unable to forward port because pod is not running. Current status=%v", pod.Status.Phase)
+	}
+
+	req := p.RESTClient.Post().
+		Resource("pods").
+		Namespace(p.NewNamespaceName).
+		Name(pod.Name).
+		SubResource("portforward")
+
+	// convert ports from []int to []string
+	ports := make([]string, 2)
+	for i, port := range requiredPorts {
+		ports[i] = strconv.Itoa(port)
+	}
+
+	return p.forwardPorts("POST", req.URL(), ports)
+
+}
+
+func (p *Kubernetesp) forwardPorts(method string, url *url.URL, requiredPorts []string) error {
+	transport, upgrader, err := spdy.RoundTripperFor(p.clusterConfig)
+	if err != nil {
+		return err
+	}
+	dialer := spdy.NewDialer(upgrader, &http.Client{Transport: transport}, method, url)
+	fw, err := portforward.New(dialer, requiredPorts, p.StopChannel, p.ReadyChannel, p.cmdOut, p.cmdErr)
+	if err != nil {
+		return err
+	}
+
+	return fw.ForwardPorts()
+}
+
+// CopyTar copies the files defined in each filePair in files to the pod
+// provided. Called by controller when initContainer status is running.
+func (p *Kubernetesp) CopyTar(files []FilePair, pod *apiv1.Pod) error {
+	p.Debug("copyTar Called", "files", files, "pod", pod.ObjectMeta.Name)
+	//Set up new pipe
+	pipeReader, pipeWriter := io.Pipe()
+
+	var err error
+	go func() {
+		defer pipeWriter.Close()
+		tarErr := makeTar(files, pipeWriter)
+		if tarErr != nil {
+			p.Error("error writing tar", "err", tarErr)
+			err = tarErr
+		}
+	}()
+	if err != nil {
```

Comment:
This does not capture any possible error in the preceding goroutine, since you don't wait for that goroutine to finish. If we need to wait, just don't have the above in a goroutine.